File: model/PixivNovel.py
import codecs
import json
from datetime import datetime
import common.datetime_z as datetime_z

import common.PixivHelper as PixivHelper
from common.PixivException import PixivException
from model.PixivImage import PixivTagData

MAX_LIMIT = 10


class PixivNovel:

    def __init__(self, novel_id, novel_json, tzInfo=None, dateFormat=None) -> None:
        self.novel_id = self.imageId = novel_id
        self.novel_json_str = novel_json
        self.content = ""

        # compatibility
        self.artist = None
        self.artist_id = 0
        self.imageTitle = ""
        self.worksDate = ""
        self.worksDateDateTime = datetime.fromordinal(1)
        self.imageTags = None
        self.tags = None
        self.bookmark_count = 0
        self.image_response_count = 0

        # series info
        self.seriesNavData = None
        self.seriesId = 0
        self.seriesOrder = 0

        # novel specific
        self.isOriginal = False
        self.isBungei = False
        self.language = ""
        self.xRestrict = False
        self.uploadDate = datetime.fromordinal(1)

        # doesn't apply
        self.worksResolution = ""
        self.imageMode = "Novel"

        self._tzInfo = tzInfo
        self.dateFormat = dateFormat
        self.parse()

    def parse(self):
        js = json.loads(self.novel_json_str)
        if js["error"]:
            raise PixivException("Cannot get novel details",
                                 errorCode=PixivException.UNKNOWN_IMAGE_ERROR,
                                 htmlPage=self.novel_json_str)

        root = js["body"]

        self.imageTitle = root["title"]
        self.content = root["content"]
        self.artist_id = root["userId"]
        self.bookmark_count = root["bookmarkCount"]
        self.image_response_count = root["imageResponseCount"]
        self.seriesNavData = root["seriesNavData"]
        if root["seriesNavData"] is not None:
            self.seriesId = root["seriesNavData"]["seriesId"]
            self.seriesOrder = root["seriesNavData"]["order"]
        self.isOriginal = root["isOriginal"]
        self.isBungei = root["isBungei"]
        self.language = root["language"]
        self.xRestrict = root["xRestrict"]

        # datetime
        self.worksDateDateTime = datetime_z.parse_datetime(root["createDate"])
        self.uploadDate = datetime_z.parse_datetime(root["uploadDate"])
        self.js_createDate = root["createDate"]  # store for json file
        if self._tzInfo is not None:
            self.worksDateDateTime = self.worksDateDateTime.astimezone(self._tzInfo)
            self.uploadDate = self.uploadDate.astimezone(self._tzInfo)

        tempDateFormat = self.dateFormat or "%Y-%m-%d"     # 2018-07-22, else configured in config.ini
        self.worksDate = self.worksDateDateTime.strftime(tempDateFormat)

        # tags
        self.imageTags = list()
        self.tags = list()
        tags = root["tags"]
        if tags is not None:
            tags = root["tags"]["tags"]
            for tag in tags:
                self.imageTags.append(tag["tag"])
                self.tags.append(PixivTagData(tag["tag"], tag))

        # append original tag
        if root["isOriginal"]:
            self.imageTags.append("オリジナル")
            tag = {"tag": "オリジナル",
                        "locked": True,
                        "deletable": False,
                        "userId": "",
                        "romaji": "original",
                        "translation": {
                            "en": "original"
                        }
                   }
            self.tags.append(PixivTagData(tag["tag"], tag))

    def write_content(self, filename):
        ft = open("novel_template.html")
        template_str = ft.read()
        ft.close()

        fh = None
        try:
            PixivHelper.makeSubdirs(filename)
            fh = codecs.open(filename, 'wb', encoding='utf-8')
        except IOError:
            fh = codecs.open(str(self.novel_id) + ".html", 'wb', encoding='utf-8')
            PixivHelper.get_logger().exception("Error when saving novel: %s, file is saved to: %s.html", filename, str(self.novel_id))

        if fh is not None:
            content_str = template_str.replace("%title%", self.imageTitle)
            content_str = content_str.replace("%novel_json_str%", self.novel_json_str)
            fh.write(content_str)
            fh.close()


class NovelSeries:

    def __init__(self, series_id, series_json) -> None:
        self.series_id = series_id
        self.series_str = series_json
        self.series_list = list()
        self.series_list_str = dict()
        self.total = 0
        self.series_name = ""

        self.parse()

    def parse(self):
        js = json.loads(self.series_str)
        if js["error"]:
            raise PixivException("Cannot get novel series content details",
                                 errorCode=PixivException.UNKNOWN_IMAGE_ERROR,
                                 htmlPage=self.series_str)
        # from publishedContentCount or total or displaySeriesContentCount ????
        self.total = js["body"]["total"]
        self.series_name = js["body"]["title"]

    def parse_series_content(self, page_info, current_page):
        js = json.loads(page_info)
        if js["error"]:
            raise PixivException("Cannot get novel series content details",
                                 errorCode=PixivException.UNKNOWN_IMAGE_ERROR,
                                 htmlPage=page_info)

        self.series_list.extend(js["body"]["page"]["seriesContents"])
        self.series_list_str[current_page] = page_info

File: lib/model/pixiv_novel.dart
/// PixivNovel model.
library;

import 'dart:convert';
import 'dart:io';

import '../common/datetime_z.dart' as datetime_z;
import '../common/pixiv_exception.dart';
import '../common/pixiv_helper.dart' as pixiv_helper;
import 'pixiv_image.dart';

const int MAX_LIMIT = 10;

class PixivNovel {
  final int novelId;
  int get imageId => novelId;
  final String novelJsonStr;
  String content = '';

  // compatibility
  dynamic artist;
  int artistId = 0;
  String imageTitle = '';
  String worksDate = '';
  DateTime worksDateDateTime = DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);
  List<String>? imageTags;
  List<PixivTagData>? tags;
  int bookmark_count = 0;
  int image_response_count = 0;

  // Series
  Map<String, dynamic>? seriesNavData;
  int seriesId = 0;
  int seriesOrder = 0;

  // Novel-specific
  bool isOriginal = false;
  bool isBungei = false;
  String language = '';
  bool xRestrict = false;
  DateTime uploadDate = DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);

  String worksResolution = '';
  String imageMode = 'Novel';

  Duration? _tzInfo;
  String? dateFormat;
  String? jsCreateDate;

  PixivNovel(this.novelId, this.novelJsonStr,
      {Duration? tzInfo, this.dateFormat})
      : _tzInfo = tzInfo {
    parse();
  }

  void parse() {
    final js = jsonDecode(novelJsonStr) as Map<String, dynamic>;
    if (js['error'] == true) {
      throw PixivException(
        'Cannot get novel details',
        errorCode: PixivException.UNKNOWN_IMAGE_ERROR,
        htmlPage: novelJsonStr,
      );
    }
    final root = js['body'] as Map<String, dynamic>;
    imageTitle = '${root['title']}';
    content = '${root['content']}';
    artistId = int.parse('${root['userId']}');
    bookmark_count = (root['bookmarkCount'] as num? ?? 0).toInt();
    image_response_count = (root['imageResponseCount'] as num? ?? 0).toInt();
    seriesNavData = root['seriesNavData'] as Map<String, dynamic>?;
    if (seriesNavData != null) {
      seriesId = int.parse('${seriesNavData!['seriesId']}');
      seriesOrder = int.parse('${seriesNavData!['order']}');
    }
    isOriginal = root['isOriginal'] as bool? ?? false;
    isBungei = root['isBungei'] as bool? ?? false;
    language = root['language'] as String? ?? '';
    xRestrict = (root['xRestrict'] as bool?) ??
        ((root['xRestrict'] as num?)?.toInt() == 1);

    worksDateDateTime =
        datetime_z.parseDatetime(root['createDate'] as String) ?? worksDateDateTime;
    uploadDate =
        datetime_z.parseDatetime(root['uploadDate'] as String) ?? uploadDate;
    jsCreateDate = root['createDate'] as String?;
    if (_tzInfo != null) {
      worksDateDateTime = worksDateDateTime.add(_tzInfo!);
      uploadDate = uploadDate.add(_tzInfo!);
    }
    final fmt = dateFormat ?? '%Y-%m-%d';
    worksDate = pixiv_helper.strftime(fmt, worksDateDateTime);

    imageTags = [];
    tags = [];
    if (root['tags'] is Map && root['tags']['tags'] is List) {
      for (final tag in root['tags']['tags'] as List) {
        imageTags!.add('${tag['tag']}');
        tags!.add(PixivTagData('${tag['tag']}', tag));
      }
    }

    if (isOriginal) {
      imageTags!.add('オリジナル');
      final orig = {
        'tag': 'オリジナル',
        'romaji': 'original',
        'translation': {'en': 'original'},
      };
      tags!.add(PixivTagData(orig['tag'] as String, orig));
    }
  }

  Future<void> writeContent(String filename) async {
    final templateFile = File('novel_template.html');
    String template;
    try {
      template = await templateFile.readAsString();
    } catch (_) {
      template = '<html><body>%novel_json_str%</body></html>';
    }
    var contentStr = template.replaceAll('%title%', imageTitle);
    contentStr = contentStr.replaceAll('%novel_json_str%', novelJsonStr);
    try {
      await pixiv_helper.makeSubdirs(filename);
      await File(filename).writeAsString(contentStr);
    } on FileSystemException {
      await File('$novelId.html').writeAsString(contentStr);
    }
  }
}

class NovelSeries {
  final int seriesId;
  final String seriesStr;
  List<dynamic> seriesList = [];
  Map<int, String> seriesListStr = {};
  int total = 0;
  String seriesName = '';

  NovelSeries(this.seriesId, this.seriesStr) {
    parse();
  }

  void parse() {
    final js = jsonDecode(seriesStr) as Map<String, dynamic>;
    if (js['error'] == true) {
      throw PixivException(
        'Cannot get novel series content details',
        errorCode: PixivException.UNKNOWN_IMAGE_ERROR,
        htmlPage: seriesStr,
      );
    }
    total = (js['body']['total'] as num).toInt();
    seriesName = '${js['body']['title']}';
  }

  void parseSeriesContent(String pageInfo, int currentPage) {
    final js = jsonDecode(pageInfo) as Map<String, dynamic>;
    if (js['error'] == true) {
      throw PixivException(
        'Cannot get novel series content details',
        errorCode: PixivException.UNKNOWN_IMAGE_ERROR,
        htmlPage: pageInfo,
      );
    }
    seriesList.addAll(js['body']['page']['seriesContents'] as List);
    seriesListStr[currentPage] = pageInfo;
  }
}
